cuML/src/solver/sgd.h

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "ml_utils.h"
+#include <cuda_utils.h>
+#include <linalg/gemv.h>
+#include <stats/mean.h>
+#include <stats/mean_center.h>
+#include <linalg/add.h>
+#include <linalg/subtract.h>
+#include <linalg/norm.h>
+#include <linalg/eltwise.h>
+#include <linalg/unary_op.h>
+#include <linalg/cublas_wrappers.h>
+#include <matrix/math.h>
+#include <matrix/matrix.h>
+#include "glm/preprocess.h"
+#include "shuffle.h"
+#include <functions/linearReg.h>
+#include <functions/logisticReg.h>
+#include <functions/hinge.h>
+#include "learning_rate.h"
+
+namespace ML {
+namespace Solver {
+
+using namespace MLCommon;
+
+template<typename math_t>
+void sgdFit(math_t *input,
+		    int n_rows,
+		    int n_cols,
+		    math_t *labels,
+		    math_t *coef,
+		    math_t *intercept,
+		    bool fit_intercept,
+		    int batch_size,
+		    int epochs,
+		    ML::lr_type lr_type,
+		    math_t eta0,
+		    math_t power_t,
+		    ML::loss_funct loss,
+		    Functions::penalty penalty,
+		    math_t alpha,
+		    math_t l1_ratio,
+		    bool shuffle,
+		    math_t tol,
+		    int n_iter_no_change,
+		    cublasHandle_t cublas_handle,
+		    cusolverDnHandle_t cusolver_handle) {
+
+	ASSERT(n_cols > 1,
```

Comment:
Why is it not allowed to have a single feature?

Reply:
We can allow a single feature. But, usually it's slow on GPUs.